### PR TITLE
WIP Cohere integration

### PR DESCRIPTION
**Pull request by jamescalam**
## Problem

Continuation of #57 (@DosticJelena is out until weds). Covers implementation of cohere encoder, I am adding tests and dependencies.

## Solution

Adding unit tests and dependencies

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Infrastructure change (CI configs, etc)
- [ ] Non-code change (docs, etc)
- [ ] None of the above: (explain here)

## Test Plan

Adding unit tests

### CODE REVIEW

File: pinecone_text/dense/cohere_encoder.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        :param model_name: The name of the embedding model to use. See https://docs.cohere.com/reference/embed
+        :param kwargs: Additional arguments to pass to the underlying cohere client. See https://github.com/openai/openai-python
+        """
+        if not _cohere_installed:
+            raise ImportError(
+                "Failed to import cohere. Make sure you install cohere extra "
+                "dependencies by running: "
+                "`pip install pinecone-text[cohere]"
+            )
+        if model_name not in CohereEncoderName.list_models():
```
Comment (acatav) [inline pinecone_text/dense/cohere_encoder.py:61]: I'm fine with letting Cohere API to handle this assertion, but we can also keep this assertion here

File: pinecone_text/dense/cohere_encoder.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    SEARCH_QUERY = "search_query"
+    # CLASSIFICATION = "classification"
+    # CLUSTERING = "clustering"
+
+
+class CohereEncoder(BaseDenseEncoder):
+    """
+    Cohere's text embedding wrapper. See https://docs.cohere.com/reference/embed
+
+    Note: You should provide an API key as the environment variable COHERE_API_KEY.
```
Comment (acatav) [inline pinecone_text/dense/cohere_encoder.py:39]:
```suggestion
    Note: You should provide an API key as the environment variable `CO_API_KEY`.
```

File: tests/unit/test_cohere_encoder.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+import pytest
```
Comment (acatav) [inline tests/unit/test_cohere_encoder.py:1]: Can you please add a system test similar to the one we have for openai? (it's under `tests/system`)

**Review by acatav — CHANGES_REQUESTED** [review]

File: tests/unit/test_cohere_encoder.py (hunk 3 above)
Comment (acatav) [reply tests/unit/test_cohere_encoder.py:1]: I already put a secret on github under  `CO_API_KEY` env var, so the test can simply pass `None` to the encoder in the api key

File: pinecone_text/dense/cohere_encoder.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
+from enum import Enum, StrEnum
```
[the remark this answers is not in the document]
Comment (jamescalam) [reply pinecone_text/dense/cohere_encoder.py:1]: I didn't know this — thanks will switch to `Enum`!

File: pinecone_text/dense/cohere_encoder.py (hunk 1 above)
Comment (jamescalam) [reply pinecone_text/dense/cohere_encoder.py:61]: yes I wanted to add in case someone tries to use `v2` model that doesn't include `input_type` params

**Review by acatav — APPROVED** [review]
LGTM. Just follow my minor comments and it's ready for merge :)

**Review by acatav — CHANGES_REQUESTED** [review]

**Review by acatav — CHANGES_REQUESTED** [review]
LGTM. just follow the minor comments in if all checks are pass we are good to go :)

**Review by acatav — APPROVED** [review]

**Pull request merged**